skills_ml/algorithms/preprocessing/__init__.py

```diff
@@ -0,0 +1,82 @@
+from functools import reduce
+
+from skills_ml.algorithms.string_cleaners.nlp import NLPTransforms, deep
+
+
+class NLPPipeline(object):
+    """A simple nlp preprocessing pipeline.
+
+    This class will compose preprocessing functions together to be passed to different stages(training/prediction)
+    to assert the same preprocessing procedrues.
+
+    Example:
+        ```python
+        jp = JobPostingCollectionSample()
+        pipe = NLPPipeline(
+            partial(fields_joiner, document_schema_fields=['description']),
+            html_cleaner,
+            sentence_tokenizer,
+            str_cleaner,
+            word_tokenizer
+        )
+        result = pipe.run(jp)
+        ```
+
+    Attributes:
+        functions (func): a series of generator functions that takes another generator as input
+
+    """
+    def __init__(self, *functions):
+        self.functions = list(functions)
+
+    @property
+    def compose(self):
+        """compose functions
+
+        Returns:
+            generator: a generator objet which is not materialized yet
+        """
+        return reduce(lambda f, g: lambda x: g(f(x)), self.functions, lambda x: x)
+
+    def run(self, generator):
+        """materialize the generator object
+
+        Returns:
+            list: a list of itmes after apply all the functions on them
+        """
+        return list(self.compose(generator))
+
+    @property
+    def description(self):
+        """pipeline description"""
+        return [f.__doc__ for f in self.functions]
+
+
+def fields_joiner(generator, document_schema_fields=['description','experienceRequirements', 'qualifications', 'skills']):
+    """Join selected fields. Each item is a document in json format."""
+    for document in generator:
```

Comment:
Is there any reason why each individual function needs to iterate through the generator? Could the functions just take in the atomic units and have the Pipeline handle all of the iteration? Then you could have type annotations that more directly define the data type that the function is expecting.